File: libws/src/enc_util.js
// web/enc_util.js

// Converts ArrayBuffer to Base64 string
function arrayBufferToBase64(buffer) {
    let binary = '';
    const bytes = new Uint8Array(buffer);
    bytes.forEach(b => binary += String.fromCharCode(b));
    return window.btoa(binary);
}

// Converts Base64 string to ArrayBuffer
function base64ToArrayBuffer(base64) {
    const binary = window.atob(base64);
    const buffer = new Uint8Array(binary.length);
    for (let i = 0; i < binary.length; i++) {
        buffer[i] = binary.charCodeAt(i);
    }
    return buffer.buffer;
}

// Generates an ephemeral X25519 key pair
export async function generateKeypair() {
    const keys = await window.crypto.subtle.generateKey(
        {
            name: "ECDH",
            namedCurve: "X25519"
        },
        true,
        ["deriveKey", "deriveBits"]
    );
    return keys;
}

// Exports a public key to base64
export async function exportPublicKey(publicKey) {
    const raw = await window.crypto.subtle.exportKey("raw", publicKey);
    return arrayBufferToBase64(raw);
}

// Imports a public key from base64
export async function importPublicKey(base64) {
    const raw = base64ToArrayBuffer(base64);
    return window.crypto.subtle.importKey(
        "raw",
        raw,
        { name: "ECDH", namedCurve: "X25519" },
        true,
        []
    );
}

// Derives a shared symmetric AES-GCM key
export async function deriveSharedKey(privateKey, peerPublicKey) {
    const key = await window.crypto.subtle.deriveKey(
        {
            name: "ECDH",
            public: peerPublicKey
        },
        privateKey,
        {
            name: "AES-GCM",
            length: 256
        },
        false,
        ["encrypt", "decrypt"]
    );
    return key;
}

// Encrypts payload using AES-GCM
export async function encryptPayload(symmetricKey, plaintext) {
    const enc = new TextEncoder();
    const iv = window.crypto.getRandomValues(new Uint8Array(12));
    const ciphertext = await window.crypto.subtle.encrypt(
        {
            name: "AES-GCM",
            iv: iv
        },
        symmetricKey,
        enc.encode(plaintext)
    );
    const combined = new Uint8Array(iv.length + ciphertext.byteLength);
    combined.set(iv, 0);
    combined.set(new Uint8Array(ciphertext), iv.length);
    return combined;
}

// Decrypts payload using AES-GCM
export async function decryptPayload(symmetricKey, combined) {
    const iv = combined.slice(0, 12);
    const ciphertext = combined.slice(12);
    const decrypted = await window.crypto.subtle.decrypt(
        {
            name: "AES-GCM",
            iv: iv
        },
        symmetricKey,
        ciphertext
    );
    const dec = new TextDecoder();
    return dec.decode(decrypted);
}

File: libws/src/enc_util.rs
// src/enc_util.rs

use x25519_dalek::{EphemeralSecret, PublicKey as X25519PublicKey, StaticSecret};
use aes_gcm::{Aes256Gcm, aead::{Aead, KeyInit, OsRng, generic_array::GenericArray, Payload}};
use rand_core::RngCore;
use base64::{encode as b64_encode, decode as b64_decode};

/// Represents a generated key pair
pub struct KeyPair {
    pub private_key: StaticSecret,
    pub public_key: X25519PublicKey,
}

/// Represents a derived symmetric key
pub struct SymmetricKey {
    key: Aes256Gcm,
}

impl SymmetricKey {
    pub fn encrypt(&self, plaintext: &[u8]) -> Result<Vec<u8>, String> {
        let nonce = Self::generate_nonce();
        let payload = Payload { msg: plaintext, aad: &[] };
        self.key.encrypt(&nonce, payload)
            .map(|mut ciphertext| {
                // Prepend nonce to ciphertext for later decryption
                let mut combined = nonce.to_vec();
                combined.append(&mut ciphertext);
                combined
            })
            .map_err(|e| format!("Encryption failed: {:?}", e))
    }

    pub fn decrypt(&self, combined: &[u8]) -> Result<Vec<u8>, String> {
        if combined.len() < 12 {
            return Err("Invalid ciphertext".to_string());
        }
        let (nonce_bytes, ciphertext) = combined.split_at(12);
        let nonce = GenericArray::from_slice(nonce_bytes);
        self.key.decrypt(nonce, Payload { msg: ciphertext, aad: &[] })
            .map_err(|e| format!("Decryption failed: {:?}", e))
    }

    fn generate_nonce() -> GenericArray<u8, typenum::U12> {
        let mut nonce_bytes = [0u8; 12];
        OsRng.fill_bytes(&mut nonce_bytes);
        GenericArray::clone_from_slice(&nonce_bytes)
    }
}

/// Generates a new ephemeral key pair
pub fn generate_keypair() -> KeyPair {
    let private_key = StaticSecret::new(OsRng);
    let public_key = X25519PublicKey::from(&private_key);
    KeyPair { private_key, public_key }
}

/// Serializes a public key into a base64 string
pub fn serialize_public_key(public_key: &X25519PublicKey) -> String {
    b64_encode(public_key.as_bytes())
}

/// Deserializes a base64 string into a public key
pub fn deserialize_public_key(encoded: &str) -> Result<X25519PublicKey, String> {
    b64_decode(encoded)
        .map_err(|e| format!("Base64 decode failed: {:?}", e))
        .and_then(|bytes| {
            if bytes.len() != 32 {
                Err("Invalid public key length".to_string())
            } else {
                Ok(X25519PublicKey::from(bytes.as_slice()))
            }
        })
}

/// Derives a symmetric session key from our private key and peer's public key
pub fn derive_shared_key(my_private_key: &StaticSecret, peer_public_key: &X25519PublicKey) -> SymmetricKey {
    let shared_secret = my_private_key.diffie_hellman(peer_public_key);
    let shared_bytes = shared_secret.as_bytes();
    let key = Aes256Gcm::new(GenericArray::clone_from_slice(shared_bytes));
    SymmetricKey { key }
}
